src/coreclr/System.Private.CoreLib/src/System/GC.CoreCLR.cs

```diff
@@ -614,6 +614,59 @@ internal static void RegisterMemoryLoadChangeNotification(float lowMemoryPercent
             }
         }
 
+        [Flags]
+        internal enum NoGCRegionCallbackWork
+        {
+            None,
+            Execute,
+            Detach
+        }
+
+        internal enum EnableNoGCRegionCallbackStatus
+        {
+            Succeed,
+            InsufficientBudget
+        }
+
+        private static Action? _NoGCCallback;
+
+        [UnmanagedCallersOnly]
+        private static unsafe void NoGCRegionCallback(NoGCRegionCallbackWork work)
+        {
+            Action? callback = _NoGCCallback;
+            if (work.HasFlag(NoGCRegionCallbackWork.Detach))
+            {
+                _NoGCCallback = null;
+            }
+            if (work.HasFlag(NoGCRegionCallbackWork.Execute))
+            {
+                callback?.Invoke();
+            }
+        }
+
```

Comment:
@AaronRobinsonMSFT is this the recommended way to do this sort thing? we have a callback that we want to call on the finalizer thread. is it better to do it this way or just call the callback directly on the native code side?